Implement request forwarding for ZkRoutingDataWriter

NealSun96 opened the pull request:
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

Fixes #778 , #794 , #795 , #800 , #801

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

Because of the single-writer design for `ZkRoutingDataWriter`, we are utilizing `ZkDistributedLeaderElection` to choose a leader instance and non-leaders will forward write requests to the leader. This PR:
1. implements the logic for request forwarding; 
2. changes the current MSDS endpoints behaviors such that the endpoints now respect the return values of underlying methods (previously, if the underlying methods return `false`, the endpoints will still return `OK` which is not desirable); 
3. enables the previously disabled MSDS endpoints tests;
4. fixes `TestMetadataStoreDirectoryAccessor` to allow tests to be run in leader mode (the previous version disallows tests to be run in leader mode);
5. fixes 2 bugs that break endpoionts tests (#794, #795);
6. adds unit tests for follower situations in `TestZkRoutingDataWriter`, which uncovers a bug on leader election that causes null pointer for followers (#801, fixed);
7. makes ZkMetadataStoreDirectory singleton (#800);
8. added integration tests for the entire flow.


### Tests
- [x] The following is the result of the "mvn test" command on the appropriate module:
[INFO] Tests run: 131, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 24.278 s - in TestSuite
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 131, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  52.393 s
[INFO] Finished at: 2020-02-25T18:48:33-08:00
[INFO] ------------------------------------------------------------------------


### Commits

- [x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- [x] My diff has been formatted using helix-style.xml

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -77,7 +88,23 @@ public synchronized boolean addMetadataStoreRealm(String realm) {
+      if (response.getStatusLine().getStatusCode() != 201) {
+        HttpEntity respEntity = response.getEntity();
+        LOG.error("the forwarded request to leader has failed for addMetadataStoreRealm: {}",
+            respEntity != null ? EntityUtils.toString(respEntity) : "");
```
What's the point of recording ""? Do you really need the ternary operator in this case?

NealSun96 replied:
When there's no response body, it is still necessary to log "the forward request has failed". Constructing it like this reduces duplication. 
Maybe I can change "" to "No response body is returned" or something else meaningful.

narendly replied:
You could do this:

```
String errorMessage = "The forwarded request from " + (this node's hostname) + " to leader " + (leader's hostname) + " has failed for addMetadataStoreRealm()!";
if (respEntity != null) {
  errorMessage += " Response: " + EntityUtils.toString(respEntity);
}
LOG. error(errorMessage);
```

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -90,7 +117,24 @@ public synchronized boolean deleteMetadataStoreRealm(String realm) {
-    // TODO: Forward the request to leader
+    String leaderHostName = _leaderElection.getCurrentLeaderInfo().getId();
+    String url =
+        leaderHostName + "/admin/v2/namespaces/" + _namespace + "/metadata-store-realms/" + realm;
```
See above - refactor!

NealSun96 replied:
Ack.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -77,7 +88,23 @@ public synchronized boolean addMetadataStoreRealm(String realm) {
+        leaderHostName + "/admin/v2/namespaces/" + _namespace + "/metadata-store-realms/" + realm;
+    HttpPut putRequest = new HttpPut(url);
+    try {
+      HttpResponse response = _forwardHttpClient.execute(putRequest);
```
Could we configure a timeout for these requests? 

https://www.baeldung.com/httpclient-timeout

Something like 1 minute sounds reasonable. Make that 1 minute a constant somewhere and you could use it throughout.

NealSun96 replied:
I'll take a look. Very good point.

NealSun96 replied:
@narendly What's your opinion on the necessity of setting a socket timeout in addition to a connection timeout?

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -90,7 +117,24 @@ public synchronized boolean deleteMetadataStoreRealm(String realm) {
+      if (response.getStatusLine().getStatusCode() != 200) {
+        HttpEntity respEntity = response.getEntity();
+        LOG.error("the forwarded request to leader has failed for deleteMetadataStoreRealm: {}",
+            respEntity != null ? EntityUtils.toString(respEntity) : "");
```
Why is the ternary necessary?

Did you meant to have a log msg like:

"the forwarded request to leader has failed for deleteMetadataStoreRealm:  "?

NealSun96 replied:
Ditto my reply above.

narendly replied:
See above

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -136,7 +180,23 @@ public synchronized boolean addShardingKey(String realm, String shardingKey) {
+    String leaderHostName = _leaderElection.getCurrentLeaderInfo().getId();
+    String url =
+        leaderHostName + "/admin/v2/namespaces/" + _namespace + "/metadata-store-realms/" + realm
+            + "/sharding-keys/" + shardingKey;
```
refactor?

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -225,6 +302,11 @@ public synchronized boolean setRoutingData(Map<String, List<String>> routingData
+    try {
+      _forwardHttpClient.close();
+    } catch (IOException e) {
+      LOG.error("HttpClient fails to close. ", e);
```
Nit: "failed"

NealSun96 replied:
Ack.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/metadatastore/MetadataStoreDirectoryAccessor.java:
```diff
@@ -90,9 +90,13 @@ public Response getAllMetadataStoreRealms() {
     } catch (IllegalArgumentException ex) {
       return notFound(ex.getMessage());
+    } catch (IllegalStateException ex) {
+      return serverError();
```
Where are you throwing an IllegalStateException?

NealSun96 replied:
In `ZkRoutingDataWriter` if `_zkClient.isClosed()`.

pkuwm replied:
If that’s the case, I would put it in addMetadatStoreRealm or a proper place, but not here in this endpoint logic. It is not endpoint logic’s responsibility to process such exception.

narendly replied:
+1. Let's make sure the MSDS server logs the errors. it would be helpful if we could include as much detail as possible including the host name, etc.

NealSun96 replied:
If it's not handled, the exception will cause a server error anyways. Are you suggesting to handle it and not raise server error? IMO zkClient being closed is definitely a server error.

narendly replied:
You could let this IllegalStateException float up all the way to the servlet..

For instance, for uncaught exceptions, the servlet container usually catch the Exception, logs it and returns http status 500.

Just make sure MetadataStoreDirectory logs the exception properly.

NealSun96 replied:
I see. I'll remove this section of the code. Thank you both for the insight.

narendly commented on metadata-store-directory-common/src/main/java/org/apache/helix/msdcommon/constant/MetadataStoreRoutingConstants.java:
```diff
@@ -37,5 +37,6 @@
 
-
+  // The key for system properties that contains the hostname of an instance
+  public static final String HOSTNAME_SYSTEM_PROPERTY_KEY = "instance_hostname";
```
Nit: rename to MSDS_SERVER_HOSTNAME_KEY?

And for the actual value, let's use "msds_hostname" so it doesn't conflict with some general "instance_hostname" key.

NealSun96 replied:
Sure.

narendly commented on metadata-store-directory-common/src/main/java/org/apache/helix/msdcommon/constant/MetadataStoreRoutingConstants.java:
```diff
@@ -37,5 +37,6 @@
   public static final String SHARDING_KEYS = "shardingKeys";
 
-
+  // The key for system properties that contains the hostname of an instance
```
hostname of the MSDS server instance

NealSun96 replied:
Ack.

NealSun96 commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -77,7 +88,23 @@ public synchronized boolean addMetadataStoreRealm(String realm) {
-    // TODO: Forward the request to leader
+    String leaderHostName = _leaderElection.getCurrentLeaderInfo().getId();
+    String url =
+        leaderHostName + "/admin/v2/namespaces/" + _namespace + "/metadata-store-realms/" + realm;
```
Agreed.

NealSun96 replied:
`_leaderHostName` is an instance variable now since it's used during multiple occasions. I'm refactoring the calling code since it can be generalized and reduce duplication.

NealSun96 replied:
On a second thought I cannot do that because `_leaderHostName` will be changing. Modified my planned code structure for that. Duplication can still be reduced.

pkuwm commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -62,10 +71,12 @@ public ZkRoutingDataWriter(String namespace, String zkAddress) {
-    // TODO: Fill in when Helix REST implementations are ready
-    ZNRecord myServerInfo = new ZNRecord("dummy hostname");
+    ZNRecord myServerInfo = new ZNRecord(
+        System.getProperty(MetadataStoreRoutingConstants.HOSTNAME_SYSTEM_PROPERTY_KEY));
```
What if the property is not set in system?

narendly replied:
+1. We could probably throw an IllegalStateException here if System.getProperty returns null.

NealSun96 replied:
Good point. Fixing.

pkuwm commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -77,7 +88,23 @@ public synchronized boolean addMetadataStoreRealm(String realm) {
+    HttpPut putRequest = new HttpPut(url);
+    try {
+      HttpResponse response = _forwardHttpClient.execute(putRequest);
+      if (response.getStatusLine().getStatusCode() != 201) {
```
I would use created code rather than the hard coded 201: easier to read and maintain

narendly replied:
+1. Don't we already have this as part of Helix REST? See `javax.ws.rs.core.Response;`?

NealSun96 replied:
Sure.

pkuwm commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/metadatastore/MetadataStoreDirectoryAccessor.java:
```diff
@@ -153,9 +161,13 @@ public Response getShardingKeys(@QueryParam("realm") String realm) {
     try {
-      _metadataStoreDirectory.addShardingKey(_namespace, realm, shardingKey);
+      if (!_metadataStoreDirectory.addShardingKey(_namespace, realm, shardingKey)) {
+        return serverError();
```
Can we have a message in the server error message why it fails or what the client should do? Server error is too plain.

NealSun96 replied:
There are numerous different reasons why this could happen, and one single returned message is not going to cover all cases; saying it's a server error is sufficient to express that it's a server side situation. 
On the other hand, whenever `false` is returned here, there are server logs that can help us identify the cause.

pkuwm replied:
That is for server side. Think about client side: what is response for this? If you are a user, what would you like to know?

NealSun96 replied:
Here are the situations where this condition can be invoked:
1. When the request forwarded to the leader instance returns a non-ok;
2. When the request forwarded to the leader raises an IOException;
3. When writing to ZooKeeper fails. 
All of them happen on the layer of `ZkRoutingDataWriter`. I'll need some advice on how to propagate the correct information back to the REST layer - I do not see an easy way to do this. 
At the same time, I believe it's unnecessary to let clients know the exact situation among these 3 situations, because these situations are unrelated to client input and there's nothing clients can do. All these situations are strictly internal errors and are out of the hands of clients.

pkuwm replied:
I understand the possible situations and there is no need to tell the client the real reason (and it is impossible here because the api just returns true or false). I meant, include a message suggesting what to do like “server error , please contact admin” is helpful than plain response. Imagine, what would you get if you do curl this endpoint but server error? 
Anyway, I have an issue Created to improve the response. You don’t have to do it here. Focus on the main task of the pr instead.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -77,7 +88,23 @@ public synchronized boolean addMetadataStoreRealm(String realm) {
+        return false;
+      }
+    } catch (IOException e) {
+      LOG.error("the forwarded request to leader raised an exception for addMetadataStoreRealm", e);
```
More helpful to include which host is forwarding to which leader.

NealSun96 replied:
As in the hostnames? Okay.

narendly replied:
Like this:

`String errorMessage = "The forwarded request from " + (this node's hostname) + " to leader " + (leader's hostname) + " has failed for addMetadataStoreRealm()!";
if (respEntity != null) {
  errorMessage += " Response: " + EntityUtils.toString(respEntity);
}
LOG. error(errorMessage);`

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -90,7 +117,24 @@ public synchronized boolean deleteMetadataStoreRealm(String realm) {
+      }
+    } catch (IOException e) {
+      LOG.error("the forwarded request to leader raised an exception for deleteMetadataStoreRealm",
+          e);
```
Improve logging with host name and leader host name

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/metadatastore/MetadataStoreDirectoryAccessor.java:
```diff
@@ -102,9 +106,13 @@ public Response addMetadataStoreRealm(@PathParam("realm") String realm) {
     } catch (IllegalArgumentException ex) {
       return notFound(ex.getMessage());
+    } catch (IllegalStateException ex) {
+      return serverError();
```
Remove?

narendly commented:
Note on this PR:

1. As discussed offline, let's make ZkMetadataStoreDirectory a static singleton that is shared throughout the server's process. The benefit of this is that each REST server will truly act as one entity, which is necessary because we use distributed leader election to designate a single writer. 

2. With that said, we would need to add both unit and integration tests. With an introduction of a singleton class, testing might not be so straightforward, but as discussed in person, it might be possible by creating mock classes.

Good job and let's continue adding tests and reviewing this PR :)

NealSun96 commented:
Current state of the PR: missing integration tests for request forwarding, which will be added to `TestMetadataStoreDirectoryAccessor`. Everything else are ready for review.

NealSun96 commented:
Integration tests have been added.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/ZkMetadataStoreDirectory.java:
```diff
@@ -42,49 +42,77 @@
-    _routingDataReaderMap = new HashMap<>();
-    _routingDataWriterMap = new HashMap<>();
-    _routingZkAddressMap = routingZkAddressMap;
+    _zkMetadataStoreDirectoryInstance.init(namespace, zkAddress);
```
Can you do

getInstance().init(namespace, zkAddress) here to cut down on duplicate code?

NealSun96 replied:
That is a good point. Let me try that.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -40,8 +52,11 @@
+  private final String _myHostName;
 
   public ZkRoutingDataWriter(String namespace, String zkAddress) {
+    System.out.println("writer created namespace=" + namespace + " zkAddress=" + zkAddress);
```
Remove print statements? We want to avoid this kind of stuff in production code.

NealSun96 replied:
This one escaped my own review. Fixing.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -62,10 +77,21 @@ public ZkRoutingDataWriter(String namespace, String zkAddress) {
-    // TODO: Fill in when Helix REST implementations are ready
-    ZNRecord myServerInfo = new ZNRecord("dummy hostname");
+    _myHostName = System.getProperty(MetadataStoreRoutingConstants.MSDS_SERVER_HOSTNAME_KEY);
+    if (_myHostName == null) {
```
Also include the check: `_myHostName.isEmpty()`

NealSun96 replied:
Makes sense.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -62,10 +77,21 @@ public ZkRoutingDataWriter(String namespace, String zkAddress) {
     _leaderElection = new ZkDistributedLeaderElection(_zkClient,
         MetadataStoreRoutingConstants.LEADER_ELECTION_ZNODE, myServerInfo);
+
+    int timeout = 60; // seconds
```
Let's make this a Java constant so that we could make this configurable later.

`private static final int HTTP_REQUEST_TIMEOUT = 60;`

NealSun96 replied:
Ack.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -77,8 +103,9 @@ public synchronized boolean addMetadataStoreRealm(String realm) {
 
-    // TODO: Forward the request to leader
-    return true;
+    String url_suffix = "namespaces/" + _namespace + "/metadata-store-realms/" + realm;
```
1. Generally a good idea to refactor this to a private method so that it's easier to read. For instance, create `private String getUrlSuffix()`
2. Some of these string literals have already been defined in MetadataStoreDirectoryConstants.java. Could you use the constants there please?

NealSun96 replied:
1. I wanted to include the url creation as a part of `forwardRequestToLeader`, but I do not see a good way to generalize url construction. The 4 endpoints require 2 different urls; refactoring it results in an if statement, which isn't generalization. On the other hand, leaving url construction logic in the methods also clearly indicates which urls each endpoints correspond to, which is good for maintenance. 
2. In `MetadataStoreRoutingConstants`, the constants are not quite the same as the strings in these urls. (`shardingKeys` vs `sharding-keys`, `metadataStoreRealms` vs `metadata-store-realms`)

narendly replied:
1. I do see a lot of duplication here. It seems it's possible to achieve the refactoring. Let me sync with you privately.

2. If you don't have them, you can create them. That's what the constant file is for. Please clearly note what they are used for.

NealSun96 replied:
Resolved by offline discussion. A method that joins parameters will be used for url suffix building.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -77,8 +103,9 @@ public synchronized boolean addMetadataStoreRealm(String realm) {
-    // TODO: Forward the request to leader
-    return true;
+    String url_suffix = "namespaces/" + _namespace + "/metadata-store-realms/" + realm;
+    return forwardRequestToLeader(url_suffix, "put", "addMetadataStoreRealm",
```
Same here, let's make these String constants.

I believe "addMetadataStoreRealm" should come from MetadataStoreDirectoryConstants.
"put" could be a private enum called 
```
enum RequestForwardingHttpVerbs {
    PUT,
    POST,
    DELETE
}
``` something like this.

NealSun96 replied:
I agree with this. Let me add an enum for it.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -87,11 +114,12 @@ public synchronized boolean deleteMetadataStoreRealm(String realm) {
 
-    // TODO: Forward the request to leader
-    return true;
+    String url_suffix = "namespaces/" + _namespace + "/metadata-store-realms/" + realm;
```
Refactor into a private method.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -87,11 +114,12 @@ public synchronized boolean deleteMetadataStoreRealm(String realm) {
-    // TODO: Forward the request to leader
-    return true;
+    String url_suffix = "namespaces/" + _namespace + "/metadata-store-realms/" + realm;
+    return forwardRequestToLeader(url_suffix, "delete", "deleteMetadataStoreRealm",
```
Use constants. No magic string literals floating around the code.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -100,44 +128,14 @@ public synchronized boolean addShardingKey(String realm, String shardingKey) {
+        "namespaces/" + _namespace + "/metadata-store-realms/" + realm + "/sharding-keys/"
+            + shardingKey;
+    return forwardRequestToLeader(url_suffix, "put", "addShardingKey",
+        Response.Status.CREATED.getStatusCode());
```
Replace string literals.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -146,31 +144,14 @@ public synchronized boolean deleteShardingKey(String realm, String shardingKey)
+        "namespaces/" + _namespace + "/metadata-store-realms/" + realm + "/sharding-keys/"
+            + shardingKey;
+    return forwardRequestToLeader(url_suffix, "delete", "deleteShardingKey",
+        Response.Status.OK.getStatusCode());
```
String literals.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -249,4 +235,119 @@ private boolean createZkRealm(String realm) {
+        break;
+      case "delete":
+        request = new HttpDelete(url);
+        break;
```
Use an enum here instead of String literals.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -249,4 +235,119 @@ private boolean createZkRealm(String realm) {
+            "The forwarded request to leader has failed for " + endPoint + ". Current hostname: "
+                + _myHostName + " Leader hostname: " + leaderHostName;
+        if (respEntity != null) {
+          errorLog += " Response: " + EntityUtils.toString(respEntity);
```
Also a good idea to log the response code.

NealSun96 replied:
Added the response code on line 336 because it always exists. Thank you for the suggestion.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -249,4 +235,119 @@ private boolean createZkRealm(String realm) {
+          endPoint, _myHostName, leaderHostName, e);
+      return false;
+    }
+
```
Remove empty line?

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/server/ServerContext.java:
```diff
@@ -64,6 +66,8 @@ public ServerContext(String zkAddr) {
     // cannot be started correctly.
     _helixDataAccessorPool = new HashMap<>();
     _taskDriverPool = new HashMap<>();
+    // Initiate the singleton ZkMetadataStoreDirectory instance to allow it to be closed later
```
Nit: "Initialize" would the the term.

NealSun96 replied:
I agree.

NealSun96 replied:
Hm, I think my eyes were caught by the "initiate" in the comment on line 63, therefore I wrote "initiate" too.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/metadatastore/MetadataStoreDirectoryAccessor.java:
```diff
@@ -152,8 +155,13 @@ public Response getShardingKeys(@QueryParam("realm") String realm) {
       @PathParam("sharding-key") String shardingKey) {
+    if (shardingKey.charAt(0) != '/') {
+      shardingKey = "/" + shardingKey;
+    }
```
Could you add some comments about the Jersey behavior here for reference as a block comment?

NealSun96 replied:
Fixed together with the later occurrence.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/metadatastore/MetadataStoreDirectoryAccessor.java:
```diff
@@ -165,8 +173,13 @@ public Response addShardingKey(@PathParam("realm") String realm,
       @PathParam("sharding-key") String shardingKey) {
+    if (shardingKey.charAt(0) != '/') {
+      shardingKey = "/" + shardingKey;
+    }
```
Now this is duplicate code. Could you refactor this out to a private method? You could explain the JAX-RS behavior around slashes there.

NealSun96 replied:
Done so.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/mock/MockMetadataStoreDirectoryAccessor.java:
```diff
@@ -0,0 +1,118 @@
+
+/**
+ * An accessor that mocks the MetadataStoreDirectoryAccessor for testing purpose.
+ */
```
Also add here that the reason we are not putting this in the test package is because the Jersey framework does not recognize resources in the test package when it spins up servers.

NealSun96 replied:
This might not be true. On Monday we assumed that Jersey isn't able to recognize it in the test package, and we moved this file into the same package as `MetadataStoreDirectoryAccessor` for it to work, hence our conclusion. However, on Tuesday, after moving it to its own package (not in the test package, but its own package), it was not working either. 
After research, I realized that its package path was not correctly added. It was working on Monday because the package for `MetadataStoreDirectoryAccessor` was added already, causing this file to be added as well. There were some major changes to `MockHelixRestServer` for it to work in this current package. Therefore, the conclusion on Monday is not true anymore - the reason it wasn't working might not be due to the test package. 
Regardless, I think it's fine to leave this file here. However, if you believe this file is better in the test package, I can experiment with that and verify our conclusion.

narendly replied:
Why wasn't it working then? I'd like to know the reason. If you don't think putting this class in the test package causes a problem, should we move it back to the test package?

NealSun96 replied:
It wasn't working because the file's package was never added properly. Our method of overriding silently failed. 
Sure, let me experiment with moving it to a test package and let's see if it still works.

NealSun96 replied:
The experiment was successful. It shows that it works in the test package too. Now the file sits in `test.org.apache.helix.rest.server.mock`.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/mock/MockMetadataStoreDirectoryAccessor.java:
```diff
@@ -0,0 +1,118 @@
+public class MockMetadataStoreDirectoryAccessor extends MetadataStoreDirectoryAccessor {
+  // A flag that will be modified if the underlying MockZkRoutingDataWriter makes an operation
+  // against ZooKeeper
+  public static boolean operatedOnZk = false;
```
Static boolean -> make it all caps?

NealSun96 replied:
This is not a constant variable, though. It's a flag value that's used to identify if there were zookeeper operations. Should it still be all caps?

narendly replied:
I'm actually fine either way since this is just a global created for testing.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/mock/MockMetadataStoreDirectoryAccessor.java:
```diff
@@ -0,0 +1,118 @@
+      _metadataStoreDirectory = new MockZkMetadataStoreDirectory(namespace, address);
+      _mockMSDInstance = _metadataStoreDirectory;
+    } catch (InvalidRoutingDataException e) {
+      e.printStackTrace();
```
Do not leave any stacktraces here. Let's use LOG.error(...)

NealSun96 replied:
Ack.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/mock/MockMetadataStoreDirectoryAccessor.java:
```diff
@@ -0,0 +1,118 @@
+
+  /**
+   * Used to artificially create another instance of ZkMetadataStoreDirectory.
+   * ZkMetadataStoreDirectory being a singleton makes it impossible to test it,
```
Not impossible, but difficult. You're making it happen here ;)

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/mock/MockMetadataStoreDirectoryAccessor.java:
```diff
@@ -0,0 +1,118 @@
+ * An accessor that mocks the MetadataStoreDirectoryAccessor for testing purpose.
+ */
+@Path("/mock")
+public class MockMetadataStoreDirectoryAccessor extends MetadataStoreDirectoryAccessor {
```
Could you create an issue and link that issue here as a TODO? The TODO should talk about how you should make this a configurable thing where developers could just feed in configs (like # of servers and names of resources to be included) to start multiple mock server instances. The purpose is to test features that involve multiple servers (like distributed leader election, distributed locking, etc.)

NealSun96 replied:
https://github.com/apache/helix/issues/816 I'll link this in the code.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/server/resources/mock/MockMetadataStoreDirectoryAccessor.java:
```diff
@@ -0,0 +1,118 @@
+    @Override
+    public void close() {
+      _routingDataReaderMap.values().forEach(MetadataStoreRoutingDataReader::close);
+      _routingDataWriterMap.values().forEach(MetadataStoreRoutingDataWriter::close);
```
Nit: Should you clear the maps?

NealSun96 replied:
It's not necessary because MockMSD is not a singleton. A new instance is created each time.

narendly replied:
OK.

narendly commented on helix-rest/src/test/java/org/apache/helix/rest/metadatastore/accessor/TestZkRoutingDataWriter.java:
```diff
@@ -40,15 +41,35 @@
   private MetadataStoreRoutingDataWriter _zkRoutingDataWriter;
 
+  // MockWriter is used for testing request forwarding features in non-leader situations
+  class MockWriter extends ZkRoutingDataWriter {
```
Nit: MockZkRoutingDataWriter so there's no ambiguity?

NealSun96 replied:
Ditto below. This mock writer is for unit testing `ZkRoutingDataWriter`, it's not the one for integration testing. Calling this `MockZkRoutingDataWriter` actually makes it ambiguous because that would create two classes of `MockZkRoutingDataWriter`.

narendly replied:
OK

narendly commented on helix-rest/src/test/java/org/apache/helix/rest/metadatastore/accessor/TestZkRoutingDataWriter.java:
```diff
@@ -40,15 +41,35 @@
+    }
+
+    // This method does not call super() because the http call should not be actually made
+    @Override
```
Shouldn't you make the HTTP call? Otherwise you won't be covering distributed leader election? I think we **should** make the HTTP call here to the leader. Otherwise it's not a true integration test.

Or is this meant to be a unit test?

NealSun96 replied:
I think you got the wrong file. :) 
The `MockZkRoutingDataWriter` for integration testing is in `MockMetadataStoreDirectoryAccessor.java`, and that's where the actual HTTP calls are made to the leader. This file is for unit-testing, where only the forwarding logic is tested.

narendly replied:
Ok, so a unit test.

narendly commented on helix-rest/src/test/java/org/apache/helix/rest/server/MetadataStoreDirectoryAccessorTestBase.java:
```diff
@@ -0,0 +1,128 @@
+  }
+
+  protected Set<String> getAllMetadataStoreRealmsHelper() throws InvalidRoutingDataException {
+    return new HashSet<>(_routingDataReader.getRoutingData().keySet());
```
Why do you need to create a new HashSet out of a set? 

You could probably return Collections.unmodifiableSet(....keySet()) as well if you want to prevent modifications

NealSun96 replied:
Because in `TestMetadataStoreDirectoryAccessor` we need to modify this set. For example: 
```
Set<String> expectedRealmsSet = getAllMetadataStoreRealmsHelper();
expectedRealmsSet.add(TEST_REALM_3);
```
`keySet()` is unmodifiable. Making a copy makes it modifiable.

narendly replied:
getAllMetadataStoreRealmsHelper -> this naming is a little confusing. Consider renaming or explain what it does and why it's needed.

NealSun96 replied:
Renamed and added comments.

narendly commented on helix-rest/src/test/java/org/apache/helix/rest/server/TestMSDADistributedLeaderElection.java:
```diff
@@ -0,0 +1,225 @@
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import static org.apache.helix.msdcommon.constant.MetadataStoreRoutingConstants.LEADER_ELECTION_ZNODE;
```
Avoid static imports. Just qualify with "MetadataStoreRoutingConstants."

NealSun96 replied:
Ack.

narendly commented on helix-rest/src/test/java/org/apache/helix/rest/server/TestMSDADistributedLeaderElection.java:
```diff
@@ -0,0 +1,225 @@
+import static org.apache.helix.msdcommon.constant.MetadataStoreRoutingConstants.LEADER_ELECTION_ZNODE;
+
+
+public class TestMSDADistributedLeaderElection extends MetadataStoreDirectoryAccessorTestBase {
```
Nit: I think `TestMSDAccessorLeaderElection` sounds better.

NealSun96 replied:
Sure, saying "leader election" is sufficient.

narendly commented on helix-rest/src/test/java/org/apache/helix/rest/server/TestMSDADistributedLeaderElection.java:
```diff
@@ -0,0 +1,225 @@
+
+
+public class TestMSDADistributedLeaderElection extends MetadataStoreDirectoryAccessorTestBase {
+  private final String MOCK_URL_PREFIX = "/mock";
```
`private static final String` is how you declare a constant.

NealSun96 replied:
Oops. Fixing it.

narendly commented on helix-rest/src/test/java/org/apache/helix/rest/server/TestMSDADistributedLeaderElection.java:
```diff
@@ -0,0 +1,225 @@
+      namespaces.add(new HelixRestNamespace(TEST_NAMESPACE,
+          HelixRestNamespace.HelixMetadataStoreType.ZOOKEEPER, _zkAddrTestNS, false));
+      _mockHelixRestServer =
+          new MockHelixRestServer(namespaces, getBaseUri().getPort() + 1, getBaseUri().getPath(),
```
We see a lot of `getPort() + 1`'s here. Let's just create a new variable,

int newPort = getPort() + 1; and use newPort throughout? This will make your code more readable.

NealSun96 replied:
Sure.

narendly commented on helix-rest/src/test/java/org/apache/helix/rest/server/TestMSDADistributedLeaderElection.java:
```diff
@@ -0,0 +1,225 @@
+              Collections.singletonList(_auditLogger));
+      _mockHelixRestServer.start();
+    } catch (InterruptedException e) {
+      e.printStackTrace();
```
LOG.error instead of printing stacktrace please.

NealSun96 replied:
Ack.

narendly commented on helix-rest/src/test/java/org/apache/helix/rest/server/TestMSDADistributedLeaderElection.java:
```diff
@@ -0,0 +1,225 @@
+    Set<String> expectedRealmsSet = getAllMetadataStoreRealmsHelper();
+    Assert.assertFalse(expectedRealmsSet.contains(TEST_REALM_3),
+        "Metadata store directory should not have realm: " + TEST_REALM_3);
+    sendRequestAndValidate("/metadata-store-realms/" + TEST_REALM_3, "put",
```
"put" could be an enum.

NealSun96 replied:
Using the same enum as the other class.

narendly commented on helix-rest/src/test/java/org/apache/helix/rest/server/TestMSDADistributedLeaderElection.java:
```diff
@@ -0,0 +1,225 @@
+  public void testDeleteMetadataStoreRealmRequestForwarding()
+      throws InvalidRoutingDataException, IOException {
+    Set<String> expectedRealmsSet = getAllMetadataStoreRealmsHelper();
+    sendRequestAndValidate("/metadata-store-realms/" + TEST_REALM_3, "delete",
```
Use enum for "delete"

narendly commented on helix-rest/src/test/java/org/apache/helix/rest/server/TestMSDADistributedLeaderElection.java:
```diff
@@ -0,0 +1,225 @@
+        break;
+      case "delete":
+        request = new HttpDelete(url);
+        break;
```
Enum!

narendly commented:
@NealSun96 Please don't forget to rebase onto apache/zooscalability as well.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -62,10 +77,24 @@ public ZkRoutingDataWriter(String namespace, String zkAddress) {
+        .setConnectTimeout(MetadataStoreRoutingConstants.HTTP_REQUEST_FORWARDING_TIMEOUT * 1000)
+        .setConnectionRequestTimeout(
+            MetadataStoreRoutingConstants.HTTP_REQUEST_FORWARDING_TIMEOUT * 1000)
+        .setSocketTimeout(MetadataStoreRoutingConstants.HTTP_REQUEST_FORWARDING_TIMEOUT * 1000)
```
Is there a strong and compelling reason why you couldn't include * 1000 in the constant itself?

NealSun96 replied:
Good point. Let me move the multiplication.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -77,8 +106,11 @@ public synchronized boolean addMetadataStoreRealm(String realm) {
+    String urlSuffix =
+        constructUrlSuffix(MetadataStoreRoutingConstants.MSDS_GET_ALL_REALMS_ENDPOINT, realm);
+    return forwardRequestToLeader(urlSuffix,
+        MetadataStoreRoutingConstants.HttpRequestForwardingVerbs.PUT, "addMetadataStoreRealm",
```
1. If you want these HTTP verbs included in `MetadataStoreRoutingConstants `, you could use a more generalized name like `HttpVerbs`.  `HttpRequestForwardingVerbs` if you want to keep it as an enum in this class.
2. If you want to generalize and put this in MetadataStoreRoutingConstants, then I'm pretty sure you could already find the supported HTTP verb enum somewhere defined in helix-rest module. Use that instead.

NealSun96 replied:
Created `HttpConstants.RestVerbs` in `org.apache.helix.rest.common` that serves the enum purpose.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -146,31 +152,15 @@ public synchronized boolean deleteShardingKey(String realm, String shardingKey)
+        constructUrlSuffix(MetadataStoreRoutingConstants.MSDS_GET_ALL_REALMS_ENDPOINT, realm,
+            MetadataStoreRoutingConstants.MSDS_GET_ALL_SHARDING_KEYS_ENDPOINT, shardingKey);
+    return forwardRequestToLeader(urlSuffix,
+        MetadataStoreRoutingConstants.HttpRequestForwardingVerbs.DELETE, "deleteShardingKey",
```
I still see a string literal here?

NealSun96 replied:
Removing this method name parameter altogether because it's better to use url to log. The url is more verbose.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java:
```diff
@@ -249,4 +244,126 @@ private boolean createZkRealm(String realm) {
+    List<String> allUrlParameters = new ArrayList<>(
+        Arrays.asList(MetadataStoreRoutingConstants.MSDS_NAMESPACES_URL_PREFIX, _namespace));
+    allUrlParameters.addAll(Arrays.asList(urlParams));
+    // HttpUriRequest will convert all double slashes to single slashes
```
Do not rely on this behavior. Let's make sure this method creates a valid URL.

Do you think making this assumption is hacky? Could we improve it?

NealSun96 replied:
We now take out all possible double slashes.

narendly commented on metadata-store-directory-common/src/main/java/org/apache/helix/msdcommon/constant/MetadataStoreRoutingConstants.java:
```diff
@@ -20,6 +20,11 @@
  */
 
 public class MetadataStoreRoutingConstants {
+  public enum HttpRequestForwardingVerbs {
```
If you want to do this, I think you could find HTTP verbs defined elsewhere already.

narendly commented on helix-rest/src/main/java/org/apache/helix/rest/metadatastore/ZkMetadataStoreDirectory.java:
```diff
@@ -42,49 +42,69 @@
+        }
+      }
+    }
+
```
Remove empty line

NealSun96 commented:
This PR is ready to be merged, approved by @narendly
Final commit message:
## Implement request forwarding for ZkRoutingDataWriter ##
This PR added the request forwarding feature to ZkRoutingDataWriter. It also included a lot of other work, most notably: changing ZkMetadataStoreDirectory to singleton in order to allow leader forwarding, added integration tests for the request forwarding flow, modified MetadataStoreDirectoryAccessor for it to respect underlying return values, fixed numerous behavior bugs.
